### PR TITLE
Add boolean expression parser

### PR DESCRIPTION
**Pull request by SpaceKatt**
Creates a truth table from a boolean expression

### CODE REVIEW

File: .gitignore
Comment (Chris-Johnston) [inline .gitignore]: I just found this, might be a more comprehensive .gitignore

https://github.com/github/gitignore/blob/master/Global/Vim.gitignore

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: Can we alias `'~'` as NOT as well

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: Actually, if you change this to `async def bool(self, ctx, *, exp):` you don't need to wrap in quotes. It's in the discord API docs, but also can be seen here: https://github.com/Chris-Johnston/CSSBot_Py/blob/master/cogs/courseInfo.py#L59

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py:40]: Should we move this to another file? ~~(Could also make easier to test down the line if/when that's implemented)~~ I just saw that you have tests, nice!

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: Does this interpret `AB` as `A*B`?

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py:137]: Would it make more sense to throw an exception here instead?

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: Should we make these class constants as a list and then do something like `return char in boolean_operators`

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: Simplify to `if temp != '!':` ?

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: You *should* be able to do `bool(one) ^ bool(two)`

https://stackoverflow.com/questions/432842/how-do-you-get-the-logical-xor-of-two-variables-in-python

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]:
```py
>>> int(True)
1
>>> int(False)
0
```

Let's use int() instead.

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py:328]: Awesome, but maybe we should use a testing framework for this?

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py]: What about doing `str.Replace('!!', '')`, as a easy way to catch stuff like this

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [inline cogs/boolean_utils.py:312]: Would it be possible to interpret this as `A*B`?

**Review by Chris-Johnston — CHANGES_REQUESTED** [review]
Looks great, will test on my machine as well. I was picky over a few minor details.

File: .gitignore
Comment (SpaceKatt) [reply .gitignore]: Most likely, I've never run into a situation where I need more exception, but it would be good to include

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: Huh, that is pretty neat. I'll change that

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py:40]: I originally had it in another file, but I'm sick and didn't feel like dealing with creating a module structure =p

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: Yes, that could be added in pretty easily

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: No. I don't think it would be worth the effort to implement that behavior.

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py:137]: I'm unaware of how the Discord API handles exceptions and I wanted an easy way to display what was wrong with the expression. Throwing an exception would work well, I just don't know how to display that information back to a user in Discord

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: That would be a much more pythonic way to handle that

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: Sure, why not =p

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: Gotchya, I didn't know that worked like that

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py]: That sounds like a dandy idea

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py:312]: Probably, but I don't think it would be worth the effort imo

Comment (Chris-Johnston) [reply cogs/boolean_utils.py:312]: Probably could open that up as something to do "down the line" after this is merged

assuming anyone of us has time to work on it.

File: cogs/boolean_utils.py
Comment (SpaceKatt) [reply cogs/boolean_utils.py:328]: Probably, I'll have to modify the class to provide some functions to make for non-gross testing. I'm not about to write tests to match formatted strings, so probably will create a `get_bool_results()` method or something to facilitate testing

File: cogs/boolean_utils.py
Comment (Chris-Johnston) [reply cogs/boolean_utils.py:137]: iirc, the discord.py library being used catches all exceptions for you, and there's a way provided by the framework for logging or sending messages on exceptions that I'm not using.

What might be simpler is to wrap the call to this command (in the bot-specific code) with a try catch, and reply back if there was some parsing error with invalid input.

**Comment by SpaceKatt** [comment]
To do: 

- Add testing framework
- Separate classes into different files
- Change error handling to exceptions
- Implement AB as A * B

**Comment by Chris-Johnston** [comment]
Let's ship it. I'll leave it to you to get back to the todo list. (Open an issue?)

**Review by Chris-Johnston — COMMENTED** [review]
how do you even approve changes in github

**Pull request merged**
